services/storage/src/simcore_service_storage/api/rest/_files.py

```diff
@@ -0,0 +1,371 @@
+    status_code=status.HTTP_204_NO_CONTENT,
+)
+async def abort_upload_file(
+    location_id: LocationID,
+    file_id: StorageFileID,
+    query_params: Annotated[StorageQueryParamsBase, Depends()],
+    request: Request,
+):
+    dsm = get_dsm_provider(request.app).get(location_id)
+    await dsm.abort_file_upload(query_params.user_id, file_id)
+
+
+@router.post(
+    "/locations/{location_id}/files/{file_id:path}:complete",
+    response_model=Envelope[FileUploadCompleteResponse],
+    status_code=status.HTTP_202_ACCEPTED,
+)
+async def complete_upload_file(
+    query_params: Annotated[StorageQueryParamsBase, Depends()],
+    location_id: LocationID,
+    file_id: StorageFileID,
+    body: FileUploadCompletionBody,
+    request: Request,
+):
+    dsm = get_dsm_provider(request.app).get(location_id)
+    # NOTE: completing a multipart upload on AWS can take up to several minutes
+    # if it returns slow we return a 202 - Accepted, the client will have to check later
+    # for completeness
+    task = asyncio.create_task(
+        dsm.complete_file_upload(file_id, query_params.user_id, body.parts),
+        name=create_upload_completion_task_name(query_params.user_id, file_id),
+    )
+    get_completed_upload_tasks(request.app)[task.get_name()] = task
+    server_ip = socket.gethostbyname(f"{request.url.hostname}")
+    server_port = f":{request.url.port}" if request.url.port else ""
+
+    route = URL(
+        request.app.url_path_for(
+            "is_completed_upload_file",
+            location_id=f"{location_id}",
+            file_id=file_id,
+            future_id=task.get_name(),
+        )
+    ).with_query(user_id=query_params.user_id)
+    complete_task_state_url = f"{request.url.scheme}://{server_ip}{server_port}{route}"
+    response = FileUploadCompleteResponse(
+        links=FileUploadCompleteLinks(
+            state=TypeAdapter(AnyUrl).validate_python(complete_task_state_url)
+        )
+    )
+    return Envelope[FileUploadCompleteResponse](data=response)
+
+
+@router.post(
+    "/locations/{location_id}/files/{file_id:path}:complete/futures/{future_id}",
+    response_model=Envelope[FileUploadCompleteFutureResponse],
+)
+async def is_completed_upload_file(
+    query_params: Annotated[StorageQueryParamsBase, Depends()],
+    location_id: LocationID,
+    file_id: StorageFileID,
+    future_id: str,
+    request: Request,
+):
+    # NOTE: completing a multipart upload on AWS can take up to several minutes
+    # therefore we wait a bit to see if it completes fast and return a 204
+    # if it returns slow we return a 202 - Accepted, the client will have to check later
+    # for completeness
+    task_name = create_upload_completion_task_name(query_params.user_id, file_id)
+    assert task_name == urllib.parse.quote(
+        future_id
+    )  # nosec # NOTE: fastapi auto-decode path parameters
+    # first check if the task is in the app
+    if task := get_completed_upload_tasks(request.app).get(task_name):
```

Comment:
Q: let's say I now have 2 copies of storage running. This will not work.
And I don't think soemthing like sticky sessions can help. If the client is a service like the `dynamic-scheduler` it can have multiple instances making requests to storage.

Reply:
I have a similar use case in the `dynamic-scheduler`. I write the resutl of the state of a service to redis. When some client comes back and inquires about it I give it back from Redis. Also plays nice with multipe copies.

Reply:
side problem is if redis goes away.